[encoding/googlecloudlogentry] Add fields for MIG + Service metadata

andrewvc opened the pull request:
#### Description

This PR extends the VPC flow parser to emit managed instance group regions and Google service metadata (type, service name, connectivity, optional private domain) for both source and destination flows.

#### Link to tracking issue

Fixes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/44220

#### Testing

- Adds unit coverage for the new fields, introduce new test fixtures for data from these flows
- I created a new repo https://github.com/elastic/cloud-infra-repro with scripts that can configure a GCP environment that produces these logs to validate that the output here works in the real world. 


#### Documentation

The README has been updated.

andrewvc committed:
Interim work on VPC flow logs

andrewvc committed:
Fix tests

andrewvc committed:
Remove junk

andrewvc committed:
Remove junk

constanca-m commented on extension/encoding/googlecloudlogentryencodingextension/PERFORMANCE_ANALYSIS.md:
```diff
@@ -0,0 +1,169 @@
+**Memory Overhead:** For 1000 log lines, this creates:
+- 1000 ResourceLogs structures
+- 1000 ScopeLogs structures
+- Significant overhead in slice allocations and memory fragmentation
```
Hm... This was something I discussed in the beginning, and decided to go with just Resource logs (we do group them in the AWS logs encoding, for example). Can you actually prove that this is worse using a benchmark?

So, what performs better:
1. Use a resource log for every log line
2. Group by resource
   - BIG issue: the labels should be resource attributes. How are you going to make this perform well in the code (doesn't degrade CPU or memory)?

If we can't have this as a definite conclusion, then this should not be merged or shared (I know this is only a draft, but I am doubting these conclusions).

Thanks for bringing this up!

andrewvc committed:
Remove bogus perf analysis

andrewvc committed:
Cleanup extension test

andrewvc committed:
Cleanup load generation, move to terraform to keep things clean

andrewvc commented on extension/encoding/googlecloudlogentryencodingextension/extension_test.go:
```diff
@@ -196,9 +207,33 @@ func TestPayloads(t *testing.T) {
+				// NDJSON fixtures contain one JSON object per line, so we strip
+				// whitespace from each line while preserving record boundaries.
+				normalizeNDJSON(data)
+			}
```
It was not necessary, and has since been altered. Thanks!

andrewvc commented on extension/encoding/googlecloudlogentryencodingextension/extension_test.go:
```diff
@@ -29,6 +30,8 @@ func newTestExtension(t *testing.T, cfg Config) *ext {
 }
 
 func TestHandleLogLine(t *testing.T) {
+	t.Parallel()
```
I added it to the new test, felt like I may as well add it here for consistency.

andrewvc committed:
Merge branch 'main' into gcp-flowlogs-enhance

andrewvc committed:
Improve comments

andrewvc committed:
Fix linter issues

andrewvc committed:
Update test fixtures with encoding format

constanca-m commented on extension/encoding/googlecloudlogentryencodingextension/scripts/terraform/main.tf:
```diff
@@ -0,0 +1,117 @@
+resource "google_compute_network" "fixture_network" {
```
To be perfectly honest, I don't see a reason to test this end to end. I think using a unit test, via the golden package, is proof enough that it works

andrewvc replied:
I think there's room to disagree here, I'm a bit more conservative. IMHO it's a good thing to test an actual log derived from GCP where possible. Specs and implementations can differ. I think both approaches are good, they just make different risk / reward calculations.

andrewvc committed:
Remove testing infra

andrewvc committed:
Use single JSON line for golden tests, regenerate goldens, and remove NDJSON support

andrewvc commented:
Per @constanca-m 's feedback I've moved the Terraform orchestration for orchestrating a reproduction of the scenarios here out of this PR. I've moved it into its own repo here: https://github.com/elastic/cloud-infra-repro. I've also removed the NDJSON fixtures and reduced the tested data to single lines utilizing the existing multi-line JSON pattern we've used previously in this package.

andrewvc committed:
Revert unnecessary nesting in test

andrewvc committed:
Use old goldens flow

andrewvc committed:
Add chlog

constanca-m commented on extension/encoding/googlecloudlogentryencodingextension/generate_vpc_golden_test.go:
```diff
@@ -0,0 +1,77 @@
+				t.Fatalf("failed to unmarshal logs: %v", err)
+			}
+
+			if err := golden.WriteLogsToFile(filepath.Join(root, tc.outFile), logs); err != nil {
```
Ah you are right, I am out of date I guess, it only happens for metrics now: https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/golden

constanca-m reviewed (approved):
Awesome, thanks!!

andrewvc committed:
Merge branch 'main' into gcp-flowlogs-enhance

andrewvc committed:
Merge branch 'main' into gcp-flowlogs-enhance

andrewvc committed:
Merge branch 'main' into gcp-flowlogs-enhance

andrewvc committed:
Merge remote-tracking branch 'origin/main' into gcp-flowlogs-enhance

andrewvc committed:
Merge branch 'gcp-flowlogs-enhance' of github.com:andrewvc/opentelemetry-collector-contrib into gcp-flowlogs-enhance
